drivers: clock_control: stm32: clear mask bits before setting them

arbrauns opened the pull request:
Without this, setting a value of 0 leaves the bits unchanged rather than zeroing them.

arbrauns committed:
drivers: clock_control: stm32: clear mask bits before setting them

Without this, setting a value of 0 leaves the bits unchanged rather than
zeroing them.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

erwango reviewed (dismissed):
Sorry for late feedback.
Thanks for this change, @arbrauns Would you mind extend this to other stm32 clock_control files ?
Not blocking, but would be nice.

Extra nice: Adding a test case in `tests/drivers/clock_control/stm32_clock_configuration/*_devices` :-)

arbrauns committed:
tests: drivers: clock_control: stm32: clock selection with dirty registers

This makes sure clock selection works even if the registers aren't in their
default (reset) state.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns commented:
I've copied the fix to the other `stm32_clock_control_configure` definitions. I've added a test for the `common` driver, but since I don't have H7 or U5 hardware, I don't really feel confident mucking with those tests.

gautierg-st commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -253,6 +253,8 @@ static inline int stm32_clock_control_configure(const struct device *dev,
 	}
 
+	sys_clear_bits(DT_REG_ADDR(DT_NODELABEL(rcc)) + STM32_CLOCK_REG_GET(pclken->enr),
+		       STM32_CLOCK_MASK_GET(pclken->enr) << STM32_CLOCK_SHIFT_GET(pclken->enr));
```
Question: instead of having two calls, could we use the MODIFY_REG macro (specific to ST)? Or create a function that does the same job in sys_bitops.h, like sys_modify_bits?

erwango replied:
That is not related to this PR. This is non blocking

gautierg-st replied:
Completely agrees that it is not blocking. I was just wondering if a more elegant solution could be introduced, as there are probably other places where it could be useful.

gautierg-st commented on tests/drivers/clock_control/stm32_clock_configuration/stm32_common_devices/src/test_stm32_clock_configuration.c:
```diff
@@ -39,12 +39,51 @@ ZTEST(stm32_common_devices_clocks, test_sysclk_freq)
 #define STM32_I2C_DOMAIN_CLOCK_SUPPORT 0
 #endif
 
+static void i2c_set_clock(struct stm32_pclken clock)
```
You should rather pass a `const struct stm32_pclken *clock` here

erwango replied:
@arbrauns Can you take  this comment into account?
This is the only blocking comment imo

arbrauns replied:
Done.

FRASTM reviewed (commented):
Please could you also remove the second 
`#include <zephyr/sys/util.h>`
at line 18 of the drivers/clock_control/clock_stm32_l_u5.c

arbrauns committed:
drivers: clock_control: stm32: remove duplicate #include

It was being included twice. Now it's included once. CI requires a commit
description.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns commented:
I'm guessing you mean `#include <stm32_ll_utils.h>`? Sure, I can clean that up, but I don't really see how it's related to this PR.

erwango reviewed (dismissed):
LGTM, aside from the issue reported by CI
